edge/pkg/devicetwin/dtclient/device_db.go

```diff
@@ -17,20 +19,41 @@ type Device struct {
 }
 
 // SaveDevice save device
-func SaveDevice(to orm.TxOrmer, doc *Device) error {
-	num, err := to.Insert(doc)
-	klog.V(4).Infof("Insert affected Num: %d, %v", num, err)
-	return err
+func SaveDevice(o orm.Ormer, doc *Device) error {
+	err := o.DoTx(func(ctx context.Context, txOrm orm.TxOrmer) error {
+		// insert data
+		// Using txOrm to execute SQL
+		_, e := txOrm.Insert(doc)
+		// if e != nil the transaction will be rollback
+		// or it will be committed
+		return e
+	})
+	if err != nil {
+		klog.Errorf("Something wrong when insert Device data: %v", err)
+		return err
+	}
+	klog.V(4).Info("insert Device data successfully")
+	return nil
 }
 
 // DeleteDeviceByID delete device by id
-func DeleteDeviceByID(to orm.TxOrmer, id string) error {
-	num, err := to.QueryTable(DeviceTableName).Filter("id", id).Delete()
+func DeleteDeviceByID(o orm.Ormer, id string) error {
+	//num, err := o.QueryTable(DeviceTableName).Filter("id", id).Delete()
+
+	err := o.DoTx(func(ctx context.Context, txOrm orm.TxOrmer) error {
+		// insert data
+		// Using txOrm to execute SQL
+		_, e := txOrm.QueryTable(DeviceTableName).Filter("id", id).Delete()
+		// if e != nil the transaction will be rollback
+		// or it will be committed
+		return e
+	})
+
 	if err != nil {
-		klog.Errorf("Something wrong when deleting data: %v", err)
+		klog.Errorf("Something wrong when deleting Device data: %v", err)
 		return err
 	}
-	klog.V(4).Infof("Delete affected Num: %d", num)
```

Comment:
Wouldn't it be better to keep the affected num?

Reply:
In the v2 version of beego, it is officially recommended to use `o.DoTx` to handle transactions. `o.DoTx` only returns errors and no longer returns numbers as before. https://github.com/beego/beedoc/blob/master/en-US/mvc/model/transaction.md#transaction

Reply:
I get it. Does txOrm.Insert (user) return num and error? It looks like beego's example doesn't use this return parameter either.

Reply:
Yes, `txOrm.Insert` will indeed return `num` and `error`. However, `num` parameter cannot be returned in `o.DoTx` (the input parameter function defined by `o.DoTx` can only return err)